### PR TITLE
tweak some complex commands args description

### PR DESCRIPTION
**Pull request by f111fei**
PR for docs https://github.com/Microsoft/vscode-docs/pull/232

### CODE REVIEW

File: src/vs/workbench/api/node/extHostApiCommands.ts [hunk 1]
```diff
@@ -86,7 +86,8 @@ class ExtHostApiCommands {
 			description: 'Execute signature help provider.',
 			args: [
 				{ name: 'uri', description: 'Uri of a text document', constraint: URI },
-				{ name: 'position', description: 'Position in a text document', constraint: types.Position }
+				{ name: 'position', description: 'Position in a text document', constraint: types.Position },
+				{ name: 'triggerCharacter', description: 'Trigger signature help when the user types the character, like `,` or `(`', constraint: String }
 			],
```
Comment (jrieken) [inline src/vs/workbench/api/node/extHostApiCommands.ts]: this is an array and called triggerCharacters, also it's optional which I don't know if we express yet

File: src/vs/workbench/api/node/extHostApiCommands.ts [hunk 2]
```diff
@@ -101,7 +102,8 @@ class ExtHostApiCommands {
 			description: 'Execute completion item provider.',
 			args: [
 				{ name: 'uri', description: 'Uri of a text document', constraint: URI },
-				{ name: 'position', description: 'Position in a text document', constraint: types.Position }
+				{ name: 'position', description: 'Position in a text document', constraint: types.Position },
+				{ name: 'triggerCharacter', description: 'Trigger completion when the user types the character, like `,` or `(`', constraint: String }
 			],
```
Comment (jrieken) [inline src/vs/workbench/api/node/extHostApiCommands.ts]: as above

**Comment by jrieken** [comment]
Test fail because the `triggerCharacter` constraint is now mandatory

**Comment by f111fei** [comment]
@jrieken  triggerCharacter it's optional. You are right. But It`s a string. and the param is not used. See
https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/api/node/extHostLanguageFeatures.ts#L1032
and
https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/api/node/extHostLanguageFeatures.ts#L606

**Comment by jrieken** [comment]
Sorry, I was confused wrt `string` and `string[]`. 

`triggerCharacter` is used, the actual commands are defined here: https://github.com/Microsoft/vscode/blob/master/src/vs/editor/contrib/suggest/common/suggest.ts#L93 and here https://github.com/Microsoft/vscode/blob/master/src/vs/editor/contrib/parameterHints/common/parameterHints.ts#L29

**Comment by jrieken** [comment]
Currently the only way to define an optional parameter is to not provide a constraint for it. Feel free to add.

**Comment by f111fei** [comment]
@jrieken  as you say, these functions will be to the places where I said as last.

**Comment by jrieken** [comment]
Sorry, I am not good in explaining this. The character which caused a suggest or signature help request is optional. That is because both can be triggered while typing and seeing certain character or explicitly, like Cmd+Space. Hence the `triggerCharacter` is optional, but iff provided must be a `string`. Today the constraints-utils we use cannot express that notion. Therefore this PR introduces test failures (https://travis-ci.org/Microsoft/vscode/jobs/116317476#L878) because it adds the `String` constraint. Therefore it cannot be merged.

**Comment by f111fei** [comment]
yeah, `triggerCharacter` is optional. Please close the PR.  I`m sorry, Thanks!

**Comment by jrieken** [comment]
How about just removing the `constraint: String` thingie? That will make tests pass and be a valuable contribution.

**Comment by f111fei** [comment]
Good idea. Test Pass
https://github.com/Microsoft/vscode/pull/4290/commits/aeb7beb1e690f1290d112f66e2835e7a8bfd9370

**Comment by jrieken** [comment]
Thanks @f111fei

**Pull request merged**